java/client/src/main/java/glide/api/RedisClusterClient.java

```diff
@@ -808,4 +827,53 @@ public CompletableFuture<Long> sortStore(
                 concatenateArrays(new String[] {key}, sortClusterOptions.toArgs(), storeArguments);
         return commandManager.submitNewCommand(Sort, arguments, this::handleLongResponse);
     }
+
+    private static final class NativeClusterScanCursor implements ClusterScanCursor {
+        // TODO: This should be made a constant in Rust.
+        private static final String FINISHED_CURSOR_MARKER = "finished";
+
+        private String cursorHandle;
+        private boolean isFinished;
+        private boolean isClosed = false;
+
+        // This is for internal use only.
+        public NativeClusterScanCursor(String cursorHandle) {
+            this.cursorHandle = cursorHandle;
+            this.isFinished = FINISHED_CURSOR_MARKER.equals(cursorHandle);
+        }
+
+        @Override
+        public String getCursorHandle() {
+            return cursorHandle;
+        }
+
+        @Override
+        public boolean isFinished() {
+            return isFinished;
+        }
+
+        @Override
+        public void releaseCursorHandle() {
+            internalClose();
+        }
+
+        @Override
+        protected void finalize() throws Throwable {
+            try {
+                // Release the native cursor
+                this.internalClose();
+            } finally {
+                super.finalize();
+            }
+        }
+
+        private void internalClose() {
+            if (!isClosed) {
+                ClusterScanCursorResolver.releaseNativeCursor(cursorHandle);
+
+                // Mark the cursor as closed to avoid double-free (if close() gets called more than once).
```

Comment:
should we mark it as closed before we release, in case release fails?

Reply:
Makes sense.

Reply:
It would probably be good to log if there's a failure but the Rust side probably does. I'll just log anyway to be safe.

Reply:
Added logging + added a way to log exceptions to Logger.